src/services/vault/mod.rs

```diff
@@ -0,0 +1,224 @@
+            Ok(secrets.data)
+        } else {
+            self.handle_error(response).await?;
+            unreachable!()
+        }
+    }
+
+    pub async fn update_secret(
+        &self,
+        path: &str,
+        key: &str,
+        value: &str,
+    ) -> Result<bool, VaultError> {
+        let api_key = &self.get_token(false).await?;
+
+        let progress_bar = new_spinner_progress_bar();
+        progress_bar.set_message("Updating secrets... ");
+
+        let response = self
+            .vault_client
+            .update_secret(api_key, path, key, value)
+            .await?;
+
+        progress_bar.finish_and_clear();
+
+        if response.status().is_success() {
+            colored_println!("Successfully updated the secrets.");
+            response.json::<UpdateSecret>().await?;
+        } else {
+            self.handle_error(response).await?;
+        }
+
+        Ok(true)
+    }
+
+    #[async_recursion]
+    async fn get_token(&self, skip_verify: bool) -> Result<String, VaultError> {
+        debug!("Getting token ...");
+
+        let config_with_path = CLIConfig::get_config_with_path().unwrap();
+        let mut api_key = match &config_with_path.config.vault {
+            Some(vault_config) => vault_config.api_key.clone(),
+            None => {
+                debug!("No token found in config file. Prompting user to log in ...");
+                colored_println!("You are not logged in. Please log in to continue.");
+
+                self.handle_login().await?;
+                self.get_token(true).await? // Set the skip_verify flag to true on recursive call
+            }
+        };
+
+        if !skip_verify {
+            self.verify_token(&api_key).await?;
+            api_key = self.get_token(true).await?;
+        }
+
+        Ok(api_key)
+    }
+
+    async fn verify_token(&self, api_key: &str) -> Result<bool, VaultError> {
+        debug!("Verifying token ...");
+
+        let progress_bar = new_spinner_progress_bar();
+        progress_bar.set_message("Verifying the token...");
+
+        let response = self.vault_client.verify_token(api_key).await?;
+        progress_bar.finish_and_clear();
+
+        if response.status().is_success() {
+            response.json::<VerifyToken>().await?;
+            Ok(true)
+        } else {
+            self.handle_error(response).await?;
+            Ok(false)
+        }
+    }
+
+    #[async_recursion]
+    async fn handle_error(&self, response: reqwest::Response) -> Result<(), VaultError> {
+        debug!("Error: {:?}", response);
+
+        let status = response.status();
+        let message = response.text().await?;
+
+        match status {
+            StatusCode::NOT_FOUND => {
+                debug!("The requested resource was not found.");
+                return Err(VaultError::SecretNotFound);
+            }
+            StatusCode::FORBIDDEN => {
+                colored_println!("Your login session has expired/invalid. Please log in again.");
```

Comment:
for the error message, I think we should use the `CLIError` to do the output, which is in `main.rs` file

Reply:
My thinking is that since this is just a helper function, we don't make too much side effect like showing the error on to the screen. We could just return the error value and the caller can decide when and how to display the error.

For example
```rust
// in `commands/dev/config_xxx.rs`
let vault = Vault::new();
match vault.get_secrets {
    Ok(x) => { ... },
    Err(error) => { // decide whether to display error here or not }
}
```

Reply:
Meaning i dont redirect to login function if user is not logged in?

Reply:
Unless you think the error should be displayed here, and we want to do continue other actions instead of end the program. Then you should use `eprintln!()` to display on the `stderr` channel

Reply:
Hmm, that works

So we'll keep it for now with eprintln!, as we discussed since this is CLI logic we might move this whole thing away

Reply:
Ok that works too. We will figure this out on the next ticket when we're implementing the command and actually using these helper functions. Let's explore together then. For now, just change the error output to `eprintln`

Reply:
It is a good chance to learn about good API/function design.